test/cljam/io/vcf/util/validator_test.clj

```diff
@@ -0,0 +1,470 @@
+(ns cljam.io.vcf.util.validator-test
+  (:require
+   [clojure.test :refer [deftest is are testing]]
+   [cljam.io.vcf.util.validator :as v]))
+
+(deftest merge-validation-results-test
+  (are [res1 res2 expected] (= expected (#'v/merge-validation-results res1 res2))
+    nil nil nil
+    nil (#'v/error [:chr] "bad contig") {:errors {[:chr] ["bad contig"]}}
+    (#'v/error [:chr] "bad contig") nil {:errors {[:chr] ["bad contig"]}}
+
+    (#'v/error [:chr] "bad contig")
+    (#'v/error [:pos] "bad pos")
+    {:errors
+     {[:chr] ["bad contig"]
+      [:pos] ["bad pos"]}}
+
+    (#'v/error [:chr] "bad contig")
+    (#'v/error [:chr] "not defined")
+    {:errors {[:chr] ["bad contig" "not defined"]}}))
+
+(deftest prepend-keys-test
+  (are [res keys expected] (= expected (#'v/prepend-keys res keys))
+    nil [:foo] nil
+    (#'v/error [:chr] "bad contig") [:foo] {:errors {[:foo :chr] ["bad contig"]}}))
+
+(deftest prep-contig-defs-test
+  (is (= {"1" {:id "1" :length 248956422}
+          "2" {:id "2" :length 242193529}}
+         (#'v/prep-contig-defs nil
+                               [{:id "1" :length 248956422}
+                                {:id "2" :length 242193529}]))))
+
+(deftest validate-contig-test
+  (testing "file-type :vcf"
+    (let [validator (v/map->VCFValidator {})
+          prepped (#'v/prep-contig-defs validator
+                                        [{:id "1" :length 248956422}
+                                         {:id "2" :length 242193529}])
+          validator (assoc validator :defs {:contig prepped})]
+      (are [variant expected]
+           (= expected (#'v/validate-contig validator variant))
+        {:chr "1"} nil
+
+        {:chr 1}
+        {:errors
+         {[:chr]
+          [(str ":chr must be a non-empty string that consists of characters "
+                "other than whitespace, commas or angle brackets, but got 1")]}}
```

Comment:
When a string value is printed in that error message, it will be enclosed in double quotes, by which I think users can tell if the value is a string:

```clojure
(#'validator/validate-chrom nil {:chr " 1 "})
;=> {:errors
;    {[:chr]
;     [":chr must be a non-empty string that consists of characters other than whitespace, commas or angle brackets, but got \" 1 \""]}}
```

Of course, we can add the type of the value to the error message, like `"... , but got 1 (java.lang.Long)"`. But, there are many other similar error messages, and so I think it’s OK to leave them as-is for another PR until we really find them problematic.